sstables/uncompress.cc

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <cstdint>
+#include <seastar/core/fstream.hh>
+
+#include "seastar/core/future.hh"
+#include "seastar/core/temporary_buffer.hh"
+
+#include "types.hh"
+#include "exceptions.hh"
+#include "checksum_utils.hh"
+
+namespace sstables {
+
+// File data source implementation for uncompressed SSTables
+template <ChecksumUtils ChecksumType>
+class uncompressed_file_data_source_impl : public data_source_impl {
+    std::optional<input_stream<char>> _input_stream;
+    const checksum& _checksum;
+    uint64_t _underlying_pos;
+    uint64_t _pos;
+    uint64_t _beg_pos;
+    uint64_t _end_pos;
+public:
+    uncompressed_file_data_source_impl(file f, const checksum& checksum,
+    uint64_t pos, size_t len, file_input_stream_options options)
+    : _checksum(checksum)
+    , _pos(pos)
+    , _beg_pos(pos)
+    , _end_pos(pos + len)
+    {
+        // _beg_pos and _end_pos specify positions in the stream.
+        // These are not necessarily aligned on chunk boundaries.
+        // To be able to verify the checksums, we need to translate
+        // them into a range of chunks that contain the given
+        // address range, and open a file input stream to read that
+        // range. The _underlying_pos always points to the current
+        // chunk-aligned position of the file input stream.
+        auto chunk_size = checksum.chunk_size;
+        SCYLLA_ASSERT(chunk_size);
+        auto start = (_beg_pos / chunk_size) * chunk_size;
+        auto end = (_end_pos / chunk_size + 1) * chunk_size;
+        _input_stream = make_file_input_stream(std::move(f), start, end - start, std::move(options));
+        _underlying_pos = start;
+    }
+
+    virtual future<temporary_buffer<char>> get() override {
+        auto chunk_size = _checksum.chunk_size;
+        if (_pos >= _end_pos) {
+            return make_ready_future<temporary_buffer<char>>();
+        }
+        // Read the next chunk. We need to skip part of the first
+        // chunk, but then continue to read from beginning of chunks.
+        // Also, we need to take into account that the last chunk can
+        // be smaller than `chunk_size'.
+        if (_pos != _beg_pos && (_pos % chunk_size) != 0) {
```

Comment:
It's a nano-optimization, but I can't resist - better to use & (n-1) rather than the expensive %, since the chunk size has to be a power of two.

The optimization is even more worthless than usual, since typically the CPU will predict the comparison fails and move on, and execute the % operation in the background.